api/src/main/java/ai/djl/modality/Input.java

```diff
@@ -328,6 +352,113 @@ public NDList getAsNDList(NDManager manager, int index) {
+     *
+     * @return the binary encoding
+     * @throws IOException if it fails to encode part of the data
+     */
+    public byte[] encode() throws IOException {
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+            DataOutputStream os = new DataOutputStream(baos);
+            os.writeLong(serialVersionUID);
+            encodeInputBase(os);
+
+            return baos.toByteArray();
+        }
+    }
+
+    protected void encodeInputBase(DataOutputStream os) throws IOException {
+        os.writeInt(properties.size());
+        for (Entry<String, String> property : properties.entrySet()) {
+            os.writeUTF(property.getKey());
+            os.writeUTF(property.getValue());
+        }
+
+        os.writeInt(content.size());
+        for (Pair<String, BytesSupplier> c : content) {
+            if (c.getKey() != null) {
+                os.writeBoolean(true);
+                os.writeUTF(c.getKey());
+            } else {
+                os.writeBoolean(false);
+            }
+
+            byte[] cVal = c.getValue().getAsBytes();
+            os.writeInt(cVal.length);
+            os.write(cVal);
+        }
+    }
+
+    /**
+     * Decodes the input from {@link #encode()}.
+     *
+     * @param is the data to decode from
+     * @return the decoded input
+     * @throws IOException if it fails to decode part of the input
+     */
+    public static Input decode(InputStream is) throws IOException {
+        try (DataInputStream dis = new DataInputStream(is)) {
+            if (serialVersionUID != dis.readLong()) {
+                throw new IllegalArgumentException("Invalid Input version");
+            }
+
+            Input input = new Input();
+            decodeInputBase(dis, input);
+
+            return input;
+        }
+    }
+
+    protected static void decodeInputBase(DataInputStream dis, Input input) throws IOException {
+        int numProperties = dis.readInt();
+        for (int i = 0; i < numProperties; i++) {
+            String key = dis.readUTF();
+            String val = dis.readUTF();
+            input.addProperty(key, val);
+        }
+
+        int numContent = dis.readInt();
+        for (int i = 0; i < numContent; i++) {
+            boolean hasKey = dis.readBoolean();
+            String key = null;
+            if (hasKey) {
+                key = dis.readUTF();
+            }
+            int contentLength = dis.readInt();
+            byte[] contents = new byte[contentLength];
+            int contentRead = 0;
+            while (contentRead < contentLength) {
+                int newRead = dis.read(contents, contentRead, contentLength);
+                if (newRead < 0) {
+                    throw new IOException("Failed to read Input or Output content");
+                }
+                contentRead += newRead;
+            }
+            input.add(key, contents);
+        }
+    }
+
+    /** {@inheritDoc} * */
+    @Override
+    public boolean equals(Object o) {
```

Comment:
This is too heavy, we either use a id or just use Identity

Reply:
I moved it to a separate `deepEquals()` method. So, we can keep the functionality but mostly use the lighter Object equals